.github/workflows/main.yml

```diff
@@ -0,0 +1,134 @@
+name: CI Pipeline
+
+on:
+  pull_request:
+    branches:
+      - main
+    paths:
+      - 'python-flask/**'
+      - 'c-sharp-razor/**'
+      - 'rust-actix/**'
+      - 'ruby-sinatra/**'
+      - 'javascript-express/**'
+      - 'go-gin/**'
+      - 'go-gorilla/**'
+      - '/**'
+
+jobs:
+  check-root-changes:
+    runs-on: ubuntu-latest
+    outputs:
+      root_changed: ${{ steps.set-flags.outputs.root_changed }}
+      service_changed: ${{ steps.set-flags.outputs.service_changed }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Check if root files or any service changed
+        id: set-flags
+        run: |
+          # Initialize flags
+          root_changed="false"
+          any_service_changed="false"
+
+          # Get the list of changed files by comparing against the base branch
+          git fetch origin main
+          changed_files=$(git diff --name-only origin/main...HEAD)
+
+          # Check if any of the changed files are outside the known service directories
+          for file in $changed_files; do
+            if [[ ! "$file" =~ ^(python-flask/|c-sharp-razor/|rust-actix/|ruby-sinatra/|javascript-express/|go-gin/|go-gorilla/|\.github/) ]]; then
+              root_changed="true"
+              echo "Root-level file changed: $file"
+            fi
+            if [[ "$file" =~ ^(python-flask/|c-sharp-razor/|rust-actix/|ruby-sinatra/|javascript-express/|go-gin/|go-gorilla/) ]]; then
+              any_service_changed="true"
+              echo "Service-level file changed: $file"
+            fi
+          done
+
+          # Set outputs
+          echo "::set-output name=root_changed::$root_changed"
+          echo "::set-output name=service_changed::$any_service_changed"
```

Comment:
wouldn't be better to create an array of something here with the services changed already?

Reply:
Again, i tried with workflow env-vars but with no luck.

Reply:
So for time and scope i said fuck it. We already spent too much time on it imo. It runs the makefile correctly. Done deal.

Reply:
right

Reply:
In theory it is all good because it will only take a lot of time if we change many services which is very unlikely to happen! Maybe you could add just a new role if more than 2 services just test all because probably it will be faster to run the tests than set all the env again (:  

If you dont want to change it is also fair.

Reply:
I would rather leave it for now. It is necessary we can always come back and change it :)